design/20200326.extensible-certificate-controller.md

```diff
@@ -0,0 +1,388 @@
+prior to implementing this design).
+To continue to enable this, the `spec.privateKey.rotationPolicy` field
+controls how the next private key should be sourced. It supports two values:
+
+* `Always`: a new private key will be generated on every re-issuance.
+  This includes renewals as well as changes to the `spec.privateKey` and other
+  `spec` fields.
+* `Never`: private keys will never been regenerated and must be provided by the
+  user.
+
+> TODO: We may be better to have `spec.privateKey.pinnedSecretName` instead, to
+> name a Secret that contains the private key to use. We could then require
+> this key to be provided in a specific format.
+> With the `rotationPolicy` policy design, a user must pre-create the Secret
+> resource containing their private key ahead of time, which creates a conflict
+> of ownership and confusion.
+
+If the `spec.privateKey` options change during an issuance, the key will be
+regenerated and the same Secret resource will be reused to store the updated
+private key. Consumers reading this Secret (i.e. the `requestmanager`) must
+compare the public key of the named 'next private key' and if the public key
+does not match with the CertificateRequest being managed, the
+CertificateRequest should be recreated.
+
+#### requestmanager
+
+The `requestmanager` is responsible for managing a CertificateRequest resource
+for a Certificate. If a Certificate has the `Issuing` condition, it will ensure
+a CertificateRequest signed by the `status.nextPrivateKeySecretName` exists and
+matches the specification for the Certificate in `certificate.spec`.
+
+This is the controller that most closely resembles the bulk of the logic in the
+existing `certificates` controller.
+
+It will behave as follows:
+
+* If the `Issuing` condition is `True`:
+    * If the `status.nextPrivateKeySecretName` field is not set:
+        * Do nothing
+    * List all owned CertificateRequest resources, delete all that have a
+      'revision' not equal to `status.revision` + 1 (or `1` if
+      `status.revision` is not set).
+        * If multiple have the same current revision, delete any that do not 'match'
+          as per the two verification steps below. If multiple still exist, log an
+          event and delete all but one.
+    * If no CertificateRequest for the current revision exists:
+        Create a CertificateRequest based on the `certificate.spec`.
+    * Verify the public key of the certificate request matches the public key
+      of the stored `status.nextPrivateKeySecretName`
+        * If not, delete the CertificateRequest & log an Event
+    * Verify the CSR options match what is requested in `certificate.spec`
+        * If not, delete the CertificateRequest & log an Event
+* If the `Issuing` condition is `False` or not set:
+    * Delete any CertificateRequest resources that do not have a 'revision'
+      equal to `status.revision`.
+
+### issuing
```

Comment:
Where do Issuers (ACME, vanafi, self signed) come in here? most importantly external ones? Mostly thinking about advanced keys stores, does this controller still have that code built in for now and will these me moved in a later step or?

Reply:
The `requestmanager` creates a CertificateRequest and it is up to the appropriate certificaterequest controller (e.g. ACME, Venafi, etc) to then act upon the CertificateRequest.

The `issuing` controller simply copies signed `status.certificate` fields from the completed CertificateRequest, as well as the private key stored in the `status.nextPrivateKeySecretName` into the target `spec.secretName`.

> Mostly thinking about advanced keys stores, does this controller still have that code built in for now and will these me moved in a later step or?

This proposal does not (currently) intend to make it possible/define a clear means to implement a custom key store mechanism. It may be possible, however the `spec.secretName` will likely always need to be populated as we use this to perform checks/comparisons against the requested `certificate.spec` in order to determine if re-issuance is required.

Right now, if you want to persist all data in some other keystore, you'd either need a controller that 'syncs' the contents of `spec.secretName` specified on all Certificate resources into that keystore, or otherwise re-implement all the controllers discussed in this proposal (potentially borrowing code from here as you go, and replacing all 'secret access' with calls to your custom keystore)

Reply:
I'll add some notes about where external issuers fit in here to the summary/intro section, but there is no change to external issuers at all and this proposal is unrelated (https://github.com/jetstack/cert-manager/blob/master/design/20190708.certificate-request-crd.md still applies in its entirety) - the `certificates` controller is a consumer of the API described in that proposal.